design-proposals/VM-cloning.md

```diff
@@ -0,0 +1,119 @@
+* A new name for the VM (could be optional - a new name will be generated)
+* New MAC address (optional: if not provided MAC would be erased)
+* Whether to clone or skip labels / annotations etc.
+
+More parameters could be added in the future, like the new namespace for the VM, and more.
+
+## API Examples
+As said above, API can be similar to [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration).
+
+Here's a sketch for the new CRD:
+```yaml
+kind: VirtualMachineClone
+  spec:
+    source:
+      # More sources may be supported in the future, e.g. VirtualMachinInstance or VirtualMachineSnapshot
+      kind: VirtualMachine
+      name: my-other-cool-vm
+    
+    targetVirtualMachineName: my-cool-clone   # Optional - a new name can be generated, e.g. my-cool-vm-jfh54b 
+    newMacAddress: my-new-mac-address         # Optional - new mac address can be generated automatically
+    
+    # All of the fields below are optional and defaults to true
+    cloneAllAnnotations: true
+    cloneAllLabels: true
+    cloneAllDisks: true
+    cloneAllNetwork: true
+```
+
+In the future, we may want to support configuring new namespace for the cloned VM. However, this requires
+copying config-maps / secrets to another namespace which could be a potential risk.
+
+## Scalability
+I don't see any scalability issues.
+
+## Update/Rollback Compatibility
+New API so should not affect updates / rollbacks.
+
+## Implementation details / challenges (in short)
+Most of the implementation can be fairly simple, but there are a couple of things that better be in mind
+
+* Storage:
+    * Things are pretty easy thanks to CDI's [smart-cloning](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/smart-clone.md).
+With smart cloning we can simply reference the existing DVs to smart-clone them. Smart cloning is supported only for
+offline usage, therefore it is a current limitation for VM cloning.
+      * CDI would have to be installed on the cluster in order to clone disks and volumes.
+    *  Everything that is backed by a container-disk can be copied very easily, that includes CDRoms, ephemeral disks,
+  etc.
+    * Since secrets and config-maps are namespace scoped, if the VM is cloned to the same namespace it's no problem
+  to reference the same secret / config-maps in the new VM. This ia a reason for limiting the clone to be at the same
+      namespace.
+    * Host disks will not be supported.
+      
+* Network:
+  * The only important thing is the MAC address which needs to either be deleted or changed (a MAC address would be
+    generated for a new VM that does not specify it).
+    
+* Metadata:
+  * Obviously, name should be changed
+  * Also: timestamps, UIDs and similar fields should be deleted
+  * Regarding annotations / labels - I think it's better to keep them as-is by default as it's difficult to say which
+    one are important to the user. However, this can be configurable.
```

Comment:
And some may not work well with cloning. The fact that we do not know, is the risky part.

Reply:
It creates a strong dependency between new features/labels/annotations and this feature.
Such a dependency usually never ends up well. It is hard to assure it will be considered when introduced.